datadog_checks_base/datadog_checks/base/utils/discovery/discovery.py

```diff
@@ -0,0 +1,15 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from .cache import Cache
+from .filter import Filter
+
+
+class Discovery:
+    def __init__(self, get_items_func, limit=None, include=None, exclude=None, interval=None, key=None):
+        self._filter = Filter(limit, include, exclude, key)
+        self._cache = Cache(get_items_func, interval)
+
+    def get_items(self):
+        items = self._cache.get_items()
+        return self._filter.get_items(items)
```

Comment:
IIUC, we are going to filter the items even if the cache was not refreshed. This means that the filter is going to return the exact same elements it returned during in the previous call. Do you know if the filter method can take a lot of time to run? Would it be worth it to also cache this result and only filter the items if the cache got refreshed?

Reply:
That could be a problem if we modify any of the filter parameters (limit, include or exclude)

Reply:
The attributes are protected so we should not be able to do that. Modifying them would require to create a new `Discovery` instance, which would solve the issue you mention

Reply:
By protected you mean stating with `_`?

Reply:
Yep!